### PR TITLE
Add fetch improvements for app extensions.

### PR DESCRIPTION
**Pull request by dmandar**
Details at go/frc-appext-fetch-improvements
Please refer to the following commits for changes: https://github.com/firebase/firebase-ios-sdk/pull/3831/commits/8f7f9f0c8fe1e1250ed73f6da7cdc0d35de64224
https://github.com/firebase/firebase-ios-sdk/pull/3831/commits/98fb02137d23c2d7ed8c7964436ae4b015ea6cd7

### CODE REVIEW

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m [hunk 1]
```diff
@@ -101,26 +108,54 @@ - (instancetype)initWithDatabaseManager:(RCNConfigDBManager *)manager
...
+    if (options.appGroupID) {
+      _sharedUserDefaultsSuiteName = options.appGroupID;
+      /// Instantiate the shared instance.
+      _sharedUserDefaults = [self sharedUserDefaults];
+    }
   }
   return self;
 }
 
+- (NSUserDefaults *)sharedUserDefaults {
```
Comment (paulb777) [inline FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: @ryanwilson Should this be GULUserDefaults?

Comment (dmandar) [reply FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: FWIW, we use another NSUserDefaults suite currently in RC (not  GULUserDefaults).

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m [hunk 2]
```diff
@@ -32,6 +32,9 @@
 static NSString *const kRCNUserDefaultsKeyNameLastSuccessfulFetchTime = @"lastSuccessfulFetchTime";
 static const int kRCNExponentialBackoffMinimumInterval = 60 * 2;       // 2 mins.
 static const int kRCNExponentialBackoffMaximumInterval = 60 * 60 * 4;  // 4 hours.
+static NSString *const RCNExternalUserDefaultsKeyLastSuccessfulFetchTime =
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: Nit: should still have a `k` prefix to match existing constants.

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m (hunk 1 above)
Comment (ryanwilson) [reply FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: It should be `GULUserDefaults`. It uses the same underlying storage so there's no worries there, it's compatible with everything else, but when it's written to it doesn't have the bug we've seen on iOS 11 where an Apple framework tries to update UI on the thread it's fired from.

One side-effect is calling `sync` can take up to 1 second, although we need to re-evaluate that.

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m [hunk 3]
```diff
@@ -101,26 +108,54 @@ - (instancetype)initWithDatabaseManager:(RCNConfigDBManager *)manager
...
+  return sharedInstance;
+}
+
 #pragma mark - read from / update userDefaults
 - (NSString *)lastETag {
   return [_userDefaultsManager lastETag];
 }
 
 - (void)setLastETag:(NSString *)lastETag {
   [_userDefaultsManager setLastETag:lastETag];
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: The intention here then is we write to both?

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m [hunk 4]
```diff
@@ -427,7 +462,26 @@ - (BOOL)hasMinimumFetchIntervalElapsed:(NSTimeInterval)minimumFetchInterval {
...
+  if (!hasMinimumFetchIntervalElapsed && _sharedUserDefaults) {
+    // Check if the latest eTag across main app and extensions is different than our eTag. If so,
+    // let the fetch go through.
+    NSString *allAppExtensionsLatestETag =
+        [_sharedUserDefaults objectForKey:RCNExternalUserDefaultsKeyLatestETag];
+
+    // If the eTag is different, check the fetch times.
+    if (![self.lastETag isEqualToString:allAppExtensionsLatestETag]) {
+      NSNumber *latestFetchTimeAcrossAppsAndExtensions =
+          [_sharedUserDefaults objectForKey:RCNExternalUserDefaultsKeyLastSuccessfulFetchTime];
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: To confirm, if this returns `nil`, `doubleValue` should return `0` and `self.lastFetchTimeInterval` won't be set, right?

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m [hunk 5]
```diff
@@ -427,7 +462,26 @@ - (BOOL)hasMinimumFetchIntervalElapsed:(NSTimeInterval)minimumFetchInterval {
...
+      if (self.lastFetchTimeInterval < latestFetchTimeAcrossAppsAndExtensions.doubleValue) {
+        // We have an older eTag. Let the request flow through to the backend.
+        hasMinimumFetchIntervalElapsed = true;
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: `s/true/YES`

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m [hunk 6]
```diff
@@ -213,8 +213,8 @@ - (void)setUp {
     id completionBlock =
         [OCMArg invokeBlockWithArgs:_responseData[i], _URLResponse[i], [NSNull null], nil];
 
-    OCMExpect([_configFetch[i] URLSessionDataTaskWithContent:[OCMArg any]
-                                           completionHandler:completionBlock])
+    OCMStub([_configFetch[i] URLSessionDataTaskWithContent:[OCMArg any]
+                                         completionHandler:completionBlock])
         .andReturn(nil);
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Will this still call the `completionBlock`?

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m [hunk 7]
```diff
@@ -293,6 +293,49 @@ - (void)testFetchConfigsSuccessfully {
                                }];
 }
 
+- (void)testFetchConfigsWithNilCallback {
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: I'm not sure if this is named appropriately - where's the `nil` callback?

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m [hunk 8]
```diff
@@ -1164,6 +1207,95 @@ - (void)testConfigureConfigWithValidInput {
   XCTAssertEqual([config valueForKey:@"_appName"], kFIRDefaultAppName);
 }
 
+#pragma mark app extension.
+- (void)testETagAndFetchTimeUpdatedInSharedUserDefaultsForAppExtension {
+  NSMutableArray<XCTestExpectation *> *expectations =
+      [[NSMutableArray alloc] initWithCapacity:RCNTestRCNumTotalInstances];
+  for (int i = 0; i < 1; i++) {
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: This currently only runs for `i = 0`, is that expected? Or should it be `RCNTestRCNumTotalInstances`?

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m [hunk 9]
```diff
@@ -1164,6 +1207,95 @@ - (void)testConfigureConfigWithValidInput {
...
+        expectationWithDescription:
+            [NSString stringWithFormat:@"Test fetch configs for app extension - instance %d", i]];
+    XCTAssertEqual(_configInstances[i].lastFetchStatus, FIRRemoteConfigFetchStatusNoFetchYet);
+    FIRRemoteConfigFetchCompletion fetchCompletion = ^void(FIRRemoteConfigFetchStatus status,
+                                                           NSError *error) {
+      XCTAssertEqual(_configInstances[i].lastFetchStatus, FIRRemoteConfigFetchStatusSuccess);
+      XCTAssertNil(error);
+      XCTAssertNotNil(_configInstances[i].lastFetchTime);
+      XCTAssertGreaterThan(_configInstances[i].lastFetchTime.timeIntervalSince1970, 0,
+                           @"last fetch time interval should be set.");
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Nit: capital `L`

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m [hunk 10]
```diff
@@ -1164,6 +1207,95 @@ - (void)testConfigureConfigWithValidInput {
...
+  NSMutableArray<XCTestExpectation *> *expectations =
+      [[NSMutableArray alloc] initWithCapacity:RCNTestRCNumTotalInstances];
+  for (int i = 0; i < 1; i++) {
+    expectations[i] = [self
+        expectationWithDescription:
+            [NSString stringWithFormat:@"Test fetch configs for app extension - instance %d", i]];
+    XCTAssertEqual(_configInstances[i].lastFetchStatus, FIRRemoteConfigFetchStatusNoFetchYet);
+    FIRRemoteConfigFetchCompletion fetchCompletion = ^void(FIRRemoteConfigFetchStatus status,
+                                                           NSError *error) {
+      XCTAssertEqual(_configInstances[i].lastFetchStatus, FIRRemoteConfigFetchStatusSuccess);
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Nit: this is a pretty big completion handler and a lot going on here, some comments splitting things up into sections with an explanation of why it's expected would be very helpful for future readers. Note: I've only added comments and re-arranged two calls. Example:

```
// Fetch should be successful, with no errors.
XCTAssertEqual(_configInstances[i].lastFetchStatus, FIRRemoteConfigFetchStatusSuccess);
XCTAssertNil(error);

// A fetch occurred, so ensure `lastFetchTime` is valid and non-zero.
XCTAssertNotNil(_configInstances[i].lastFetchTime);
XCTAssertGreaterThan(_configInstances[i].lastFetchTime.timeIntervalSince1970, 0,
                           @"Last fetch time interval should be set.");

// The app and extension share data through UserDefaults using the `appGroupID` as the
// suite name.
NSUserDefaults *externalUserDefaults =
          [[NSUserDefaults alloc] initWithSuiteName:[self firstAppOptions].appGroupID];

// Validate the ETag is correct. 
NSString *currentETag = [externalUserDefaults objectForKey:@"frc.latestETag"];
XCTAssertTrue([currentETag isEqualToString:@"etag1-0"]);

// Validate the last fetch time is correct.
NSNumber *lastFetchTime = [externalUserDefaults objectForKey:@"frc.lastSuccessfulFetchTime"];
NSDate *lastFetchedDateTime = _configInstances[i].lastFetchTime;
XCTAssertEqual(lastFetchTime.doubleValue, [lastFetchedDateTime timeIntervalSince1970]);

// Update the eTag
XCTAssertTrue([currentETag isEqualToString:@"etag1-0"]);
[expectations[i] fulfill];
```

Comment (ryanwilson) [reply FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: (Also relevant in the second large block added below)

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m [hunk 11]
```diff
@@ -1164,6 +1207,95 @@ - (void)testConfigureConfigWithValidInput {
...
+      XCTAssertNil(error);
+      XCTAssertNotNil(_configInstances[i].lastFetchTime);
+      XCTAssertGreaterThan(_configInstances[i].lastFetchTime.timeIntervalSince1970, 0,
+                           @"last fetch time interval should be set.");
+      NSUserDefaults *externalUserDefaults =
+          [[NSUserDefaults alloc] initWithSuiteName:[self firstAppOptions].appGroupID];
+      NSString *currentETag = [externalUserDefaults objectForKey:@"frc.latestETag"];
+      NSNumber *lastFetchTime = [externalUserDefaults objectForKey:@"frc.lastSuccessfulFetchTime"];
+      XCTAssertTrue([currentETag isEqualToString:@"etag1-0"]);
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Nit: this should be a line above, right after `currentETag` is instantiated.

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m [hunk 12]
```diff
@@ -1164,6 +1207,95 @@ - (void)testConfigureConfigWithValidInput {
...
+
+  [self waitForExpectationsWithTimeout:_expectationTimeout
+                               handler:^(NSError *error) {
+                                 XCTAssertNil(error);
+                               }];
+}
+
+// An app extension should be allowed to fetch within the minimumFetchInterval if the main app/other
+// extensions have fetched updated config since.
+- (void)testFetchConfigsForAppExtension {
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: This test is pretty hard to follow. As mentioned above, more comments and some blank lines to separate unrelated blocks of code would be very helpful.

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m [hunk 13]
```diff
@@ -1164,6 +1207,95 @@ - (void)testConfigureConfigWithValidInput {
...
+  NSMutableArray<XCTestExpectation *> *expectations = [[NSMutableArray alloc] initWithCapacity:1];
+  for (int i = 0; i < 1; i++) {
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Same question re: `1`

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m (hunk 1 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: NSUserDefaults here will be called only from RC serial queue (no UI thread). Would it still be required to GULUserDefaults here?

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m (hunk 2 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: had got fb in a previous CL to not add k along with the prefix (style guide)

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m (hunk 3 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: that's right. The _userDefaultsManager writes to the existing suite (group.{bundleID}.firebase) and data there is not meant to be shared outside the process (app).

**Comment by dmandar** [comment]
> Do we also handle and test the situation where the extension updates more recently than the app, or what should we do in that situation?

This should be handled. The main app and all app extensions (only for the default namespace), all overwrite the same lastETag and fetch time fields. Any RC instance (app or extension), if otherwise hitting the cache, will check to see if the cached config has a different ETag than the latest available in sharedUserDefaults. If different, and if the last fetch time is greater than the one for the cached config, the RC instance is allowed to fetch from Remote. 
Note that all of the above logic only comes into existence if the developer has set the groupID in FIROptions.

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m (hunk 4 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: Correct. It is a no-op as long as latestFetchTimeAcrossAppsAndExtensions is same or less than the current instance's latest fetch time.

File: FirebaseRemoteConfig/Sources/RCNConfigSettings.m (hunk 5 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Sources/RCNConfigSettings.m]: Done.

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m (hunk 9 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Done.

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m (hunk 10 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Done.

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m (hunk 12 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Added more comments to test and made test code more readable.

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m (hunk 11 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Done.

File: FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m (hunk 13 above)
Comment (dmandar) [reply FirebaseRemoteConfig/Tests/Unit/RCNRemoteConfigTest.m]: Added more comments to test and made test code more readable.

**Comment by paulb777** [comment]
Closing since this would need to be reimagined anyway.
